neqo-transport/src/path.rs

```diff
@@ -18,35 +33,447 @@ use neqo_common::Datagram;
+        }
+        path.borrow_mut().probe();
+    }
+
+    /// Process elapsed time for active paths.
+    /// Returns an true if there are viable paths remaining after tidying up.
+    ///
+    /// TODO(mt) - the paths should own the RTT estimator, so they can find the PTO
+    /// for themselves.
+    pub fn process_timeout(&mut self, now: Instant, pto: Duration) -> bool {
+        let to_retire = &mut self.to_retire;
+        let mut primary_failed = false;
+        self.paths.retain(|p| {
+            if p.borrow_mut().process_timeout(now, pto) {
+                true
+            } else {
+                qdebug!([p.borrow()], "Retiring path");
+                if p.borrow().is_primary() {
+                    primary_failed = true;
+                }
+                Self::retire(to_retire, p);
+                false
+            }
+        });
+
+        if primary_failed {
+            self.primary = None;
+            // Find a valid path to fall back to.
+            if let Some(fallback) = self
+                .paths
+                .iter()
+                .rev() // More recent paths are toward the end.
+                .find(|p| p.borrow().is_valid())
+            {
+                // Need a clone as `fallback` is borrowed from `self`.
+                let path = Rc::clone(fallback);
+                qinfo!([path.borrow()], "Failing over after primary path failed");
+                let _ = self.select_primary(&path);
+                true
+            } else {
+                false
+            }
+        } else {
+            true
+        }
+    }
+
+    /// Get when the next call to `process_timeout()` should be scheduled.
+    pub fn next_timeout(&self, pto: Duration) -> Option<Instant> {
+        self.paths
+            .iter()
+            .filter_map(|p| p.borrow().next_timeout(pto))
+            .min()
+    }
+
+    /// Set the identified path to be primary.
+    /// This panics if `make_permanent` hasn't been called.
+    pub fn handle_migration(&mut self, path: &PathRef, remote: SocketAddr, now: Instant) {
+        qtrace!([self.primary().borrow()], "handle_migration");
+        for p in &self.paths {
+            qtrace!("Path {}", p.borrow());
+        }
+        // The update here needs to match the checks in `Path::received_on`.
+        // Here, we update the remote port number to match the source port on the
+        // datagram that was received.  This ensures that we send subsequent
+        // packets back to the right place.
+        path.borrow_mut().update_port(remote.port());
+
+        if path.borrow().is_primary() {
+            // Update when the path was last regarded as valid.
+            path.borrow_mut().update(now);
+            return;
+        }
+
+        if let Some(old_path) = self.select_primary(path) {
+            // Need to probe the old path if the peer migrates.
+            old_path.borrow_mut().probe();
+            // TODO(mt) - suppress probing if the path was valid within 3PTO.
```

Comment:
I'm leaving this as a TODO because until the path owns the RTT estimator, it is difficult to plumb through the necessary state.